In [ ]:
# Problem Definition for Rossmann Store Sales Forecasting

## Motivation

Rossmann là một trong những chuỗi siêu thị lớn nhất ở Đức, với hơn 3.000 cửa hàng trên toàn quốc và doanh thu hàng năm khoảng 10 tỷ Euro. Dự đoán doanh số bán hàng chính xác là rất quan trọng cho việc quản lý hiệu quả các hoạt động kinh doanh và duy trì lợi thế cạnh tranh.

### Tại sao cần dự đoán doanh số bán hàng?

1. **Quản lý hàng tồn kho (Inventory Management)**:
   - Tránh thiếu hàng: Giảm mất doanh thu từ khách hàng không tìm thấy sản phẩm
   - Tránh tồn kho quá nhiều: Giảm chi phí lưu trữ, hư hỏng, và lãng phí
   - Ước tính: Sai sót trong dự đoán có thể làm tăng chi phí lên đến 20-30%

2. **Lập kế hoạch nhân sự (Staff Planning)**:
   - Sắp xếp nhân viên phù hợp với lưu lượng khách hàng
   - Tránh chi phí nhân sự thừa hoặc thiếu
   - Tác động: Nhân sự chiếm 10-15% chi phí vận hành

3. **Chiến lược marketing và khuyến mãi (Marketing Strategies)**:
   - Đo lường hiệu quả của các chương trình khuyến mãi
   - Tối ưu hóa thời điểm và quy mô khuyến mãi
   - Ví dụ: Promo có thể tăng doanh số 10-50% tùy loại

4. **Ra quyết định tài chính (Financial Planning)**:
   - Dự báo doanh thu chính xác cho lập ngân sách
   - Quản lý dòng tiền và đầu tư
   - Giảm rủi ro tài chính từ biến động doanh số

5. **Cạnh tranh với đối thủ (Competitive Advantage)**:
   - Sử dụng dữ liệu để đưa ra quyết định nhanh hơn đối thủ
   - Tối ưu hóa giá cả và sản phẩm dựa trên dự đoán nhu cầu
   - Trong ngành bán lẻ, lợi thế cạnh tranh có thể mang lại tăng trưởng 5-10% doanh thu

### Ngữ cảnh bài toán

- **Dữ liệu lịch sử**: Từ 2013-01-01 đến 2015-07-31 cho 1.115 cửa hàng
- **Dự đoán**: Doanh số cho 6 tuần tiếp theo (từ 2015-08-01 đến 2015-09-17)
- **Yếu tố ảnh hưởng**: 
  - Thời gian: Ngày trong tuần, tháng, năm, ngày lễ
  - Khuyến mãi: Promo hàng ngày và Promo2 liên tục
  - Cạnh tranh: Khoảng cách và thời điểm mở cửa đối thủ
  - Cửa hàng: Loại cửa hàng, mức độ assortment
- **Thách thức**: Biến động theo mùa, tác động của ngày lễ, hiệu ứng khuyến mãi

### Tác động kinh doanh của dự đoán chính xác

- **Giảm chi phí**: Tối ưu inventory và nhân sự có thể tiết kiệm 5-15% chi phí
- **Tăng doanh thu**: Dự đoán tốt hơn giúp tăng hiệu quả marketing và giảm mất bán
- **Cải thiện trải nghiệm khách hàng**: Đảm bảo sản phẩm có sẵn khi cần
- **Lợi nhuận tổng thể**: Mô hình dự đoán tốt có thể cải thiện lợi nhuận 2-5%

## Evaluation Metrics

Trong bài toán dự đoán doanh số bán hàng, chúng ta cần các metrics để đánh giá độ chính xác của mô hình. Với bộ dữ liệu Rossmann, metric chính được sử dụng là **Root Mean Square Percentage Error (RMSPE)**.

### Tại sao dùng RMSPE?

- **Percentage Error**: Đo lường sai số theo tỷ lệ phần trăm, phù hợp với doanh số có thể thay đổi lớn giữa các cửa hàng.
- **Root Mean Square**: Phạt nặng các lỗi lớn hơn, khuyến khích mô hình tránh dự đoán sai quá nhiều.
- **Normalized**: Giúp so sánh giữa các cửa hàng có quy mô khác nhau.

### Công thức RMSPE

$\text{RMSPE} = \sqrt{\frac{1}{n} \sum_{i=1}^{n} \left( \frac{y_i - \hat{y}_i}{y_i} \right)^2}$

Trong đó:
- $n$: Số lượng mẫu (cửa hàng × ngày)
- $y_i$: Giá trị thực tế của doanh số (Sales)
- $\hat{y}_i$: Giá trị dự đoán
- Lưu ý: Chỉ tính cho các cửa hàng mở (Open = 1) và Sales > 0

### Ý nghĩa

- **Giá trị thấp tốt hơn**: RMSPE = 0 nghĩa là dự đoán hoàn hảo.
- **Ví dụ**: Nếu RMSPE = 0.1 (10%), nghĩa là trung bình sai số dự đoán là 10% so với doanh số thực tế.
- **Trong Kaggle**: Public leaderboard dùng 30% dữ liệu test, private dùng 70% còn lại.

### Các metrics khác có thể dùng

1. **Mean Absolute Error (MAE)**:
   $\text{MAE} = \frac{1}{n} \sum_{i=1}^{n} |y_i - \hat{y}_i|$
   - Ý nghĩa: Sai số trung bình tuyệt đối, dễ hiểu.

2. **Root Mean Square Error (RMSE)**:
   $\text{RMSE} = \sqrt{\frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2}$
   - Ý nghĩa: Sai số trung bình bình phương, phạt lỗi lớn.

3. **Mean Absolute Percentage Error (MAPE)**:
   $\text{MAPE} = \frac{100}{n} \sum_{i=1}^{n} \left| \frac{y_i - \hat{y}_i}{y_i} \right|$
   - Ý nghĩa: Tỷ lệ sai số trung bình theo phần trăm.

4. **R-squared (R²)**:
   $R^2 = 1 - \frac{\sum_{i=1}^{n} (y_i - \hat{y}_i)^2}{\sum_{i=1}^{n} (y_i - \bar{y})^2}$
   - Ý nghĩa: Tỷ lệ biến thiên được giải thích bởi mô hình (0-1, cao tốt hơn).

In [1]:
import numpy as np

# Ví dụ tính RMSPE
def rmspe(y_true, y_pred):
    """
    Root Mean Square Percentage Error
    Chỉ tính cho y_true > 0
    """
    y_true = np.array(y_true)
    y_pred = np.array(y_pred)
    mask = y_true > 0
    y_true = y_true[mask]
    y_pred = y_pred[mask]
    return np.sqrt(np.mean(((y_true - y_pred) / y_true) ** 2))

# Ví dụ dữ liệu
y_true = [100, 200, 150, 300, 250]
y_pred = [90, 210, 160, 280, 260]

rmspe_value = rmspe(y_true, y_pred)
print(f"RMSPE: {rmspe_value:.4f} ({rmspe_value*100:.2f}%)")

# So sánh với RMSE và MAE
def rmse(y_true, y_pred):
    return np.sqrt(np.mean((np.array(y_true) - np.array(y_pred)) ** 2))

def mae(y_true, y_pred):
    return np.mean(np.abs(np.array(y_true) - np.array(y_pred)))

rmse_value = rmse(y_true, y_pred)
mae_value = mae(y_true, y_pred)

print(f"RMSE: {rmse_value:.2f}")
print(f"MAE: {mae_value:.2f}")

RMSPE: 0.0678 (6.78%)
RMSE: 12.65
MAE: 12.00


### Ví dụ tính toán Metrics

Với dữ liệu mẫu:
- y_true = [100, 200, 150, 300, 250]
- y_pred = [90, 210, 160, 280, 260]

Kết quả:
- **RMSPE**: 0.0678 (6.78%) - Sai số trung bình 6.78% theo tỷ lệ
- **RMSE**: 12.65 - Sai số trung bình bình phương (đơn vị tiền tệ)
- **MAE**: 12.00 - Sai số trung bình tuyệt đối (đơn vị tiền tệ)

**Nhận xét**:
- RMSPE thấp (6.78%) cho thấy dự đoán khá chính xác
- RMSE và MAE cho thấy sai số tuyệt đối khoảng 12-13 đơn vị
- Trong thực tế với Rossmann, RMSPE mục tiêu thường dưới 15%

## Kết luận

Bài toán dự đoán doanh số bán hàng cho Rossmann là một bài toán regression quan trọng trong lĩnh vực bán lẻ. Với motivation rõ ràng về việc tối ưu hóa hoạt động kinh doanh, chúng ta cần xây dựng mô hình dự đoán chính xác sử dụng metric RMSPE.

Các bước tiếp theo:
1. Khám phá dữ liệu (EDA) - hiểu rõ các yếu tố ảnh hưởng
2. Tiền xử lý dữ liệu - xử lý missing values, feature engineering
3. Xây dựng mô hình - thử nghiệm các thuật toán ML
4. Đánh giá và tối ưu - sử dụng RMSPE để so sánh

Hiểu rõ problem definition giúp chúng ta thiết kế pipeline ML phù hợp và đạt được kết quả tốt nhất.